snorkel/classification/snorkel_classifier.py

```diff
@@ -328,10 +322,23 @@ def predict(
         prob_dict_list: Dict[str, List[torch.Tensor]] = defaultdict(list)
 
         for batch_num, (X_batch_dict, Y_batch_dict) in enumerate(dataloader):
-            prob_batch_dict = self._calculate_probs(X_batch_dict, Y_batch_dict.keys())
-            for task_name, Y in Y_batch_dict.items():
-                prob_dict_list[task_name].extend(prob_batch_dict[task_name])
-                gold_dict_list[task_name].extend(Y.cpu().numpy())
+            labels_to_tasks = self._get_labels_to_tasks(Y_batch_dict.keys())
+            prob_batch_dict = self._calculate_probs(
+                X_batch_dict, labels_to_tasks.values()
+            )
+            for label_name, Y in Y_batch_dict.items():
+                task_name = labels_to_tasks[label_name]
```

Comment:
ok so the idea here is to extract the task_name from the new format so we know how which head output to use, but then report the metric using the full metric name (with slice)?

Reply:
exactly! to de-couple the notion of label_name and task_name a bit; because we had the assumption that they always mapped 1:1, but now there might be cases where many label_names map to one task_name

Reply:
Now that I'm thinking about this, could we just supply the metadata to the dataloader separately, then alter the string representation for the printout?

Reply:
what kind of metadata? the minimum amount of information we provide is the slice labels (NxK masks for N examples and K slices)

Reply:
no i mean we set the label_name as `{base_task}:{metadata}` now to hack in the metadata. can we just pass in `{metadata}` as a separate field to the object (which then constructs the label name to print to screen) and have the label name still be `{base_task}`. that might break something else, idk

Reply:
at the moment, the abstraction requires that we copy and mask the original base_task labels to create slice labels

when training a slice model, these slice labels are used directly to compute loss, etc:
```
add_slice_labels(
    dl_valid, spam_task, S_valid, slice_names
)
```
when _evaluating_ a non-slice model, these slice_labels can be used to evaluate the primary slice_head:
```
add_slice_labels(
    dl_valid, spam_task, S_valid, slice_names, 
    include_base=False, include_ind=False
)
```

we _could_ store the metadata as a separate argument. in that case, we would not clone the base_task's labels, and we'd need to handle the training labels for the slice model to in a separate case

for speed, I'd suggest that we see if this API pattern makes sense (can share this in the context of tutorial soon)—and iterate on it afterwards